### PR TITLE
xDSL: Init Vector Dialect

### PR DESCRIPTION
**Pull request by meshtag**
This PR provides the first draft for `vector` dialect in `xDSL`. For simplicity, I have only added `vector.load` and `vector.store` operations in the dialect at the moment. 

Filecheck tests are added in `tests/filecheck/vector_ops.xdsl`. 

Corresponding `.mlir` code can be generated by running 
```
./xdsl/tools/xdsl-opt -t mlir -o tmp.mlir tests/filecheck/vector_ops.xdsl
``` 
which can be further lowered to `llvm ir` by running 
```
path-to-llvm-project/build/bin/mlir-opt --convert-scf-to-cf --convert-cf-to-llvm --convert-func-to-llvm --convert-arith-to-llvm --convert-memref-to-llvm --convert-vector-to-llvm --reconcile-unrealized-casts tmp.mlir | path-to-llvm-project/build/bin/mlir-translate --mlir-to-llvmir > tmp.ll
```

### CODE REVIEW

File: xdsl/dialects/vector.py
Comment (math-fehr) [inline xdsl/dialects/vector.py:19]: Could you change those `Exception` to `VerifyException`?

File: xdsl/dialects/vector.py
Comment (math-fehr) [inline xdsl/dialects/vector.py:18]: Could you test those lines (and the following ones, relative to verifier errors)?
You can do this directly using lit and filecheck, rather than in Python. You have an example in `arith_invalid.xdsl`.
I guess the best is to create a new `dialect/vector/` folder in `tests/filecheck/`.

**Review by math-fehr — COMMENTED** [review]
Nice!
I added a few comments, but otherwise this looks good!

File: xdsl/dialects/vector.py
Comment (meshtag) [reply xdsl/dialects/vector.py:18]: This sounds good!
I had one question though, since the program execution stops each time an exception is caught, shall I create a separate  `*vector_invalid.xdsl` file for each supposed exception? (a single file probably won't work because the execution will stop whenever the first exception is encountered)

Or is there a separate `filecheck` flag to handle such situations?

Comment (meshtag) [reply xdsl/dialects/vector.py:18]: This sounds good!
I had one question though, since the program execution stops each time an exception is caught, shall I create a separate  `*vector_invalid.xdsl` file for each supposed exception? (This sounds inelegant but a single file probably won't work because the execution will stop whenever the first exception is encountered)

Or is there a separate `filecheck` flag to handle such situations in the `filecheck` python package?
I tried finding something similar to [this](https://github.com/llvm/llvm-project/blob/e31d27e46048ccc3294d6b215dc778b3390e7834/mlir/test/Dialect/Arith/invalid.mlir#L4) directive, but couldn't find anything. The `filecheck` [roadmap ](https://github.com/mull-project/FileCheck.py/blob/main/FileCheck.pdf) doesn't seem to be providing anything useful either. 

Perhaps, the python package might implement a counterpart to `// expected-error @+1 {{}}` in future.

**Review by georgebisbas — COMMENTED** [review]
Can you not use the already use/improve/reimplement existing Store and Load from memref dialect? Is it possible?

**Comment by meshtag** [comment]
> Can you not use the already use/improve/reimplement existing Store and Load from memref dialect? Is it possible?

Expected argument types and return types are different for `vector.load/store` and `memref.load/store`. We might be able to internally use the `memref` dialect's `load/store` operations for now (by making modifications in `memref.load` and `memref.store` 's implementation to make them expect/return vector types as well and write case specific verifiers, this would probably look ugly though), but we'll have to develop separate implementation for `vector` specific operations in future (For ex. masking). 

I am of the opinion of maintaining separate implementations for `load/store` operations in `vector` and `memref` dialects, because this would make things more modular and easy to modify in future. I am open to further discussion on this topic though (if required).

**Comment by webmiche** [comment]
> Can you not use the already use/improve/reimplement existing Store and Load from memref dialect? Is it possible?

I think it is possible fundamentally, though I feel that the design @meshtag is presenting here makes more sense. I feel that separating abstractions is generally important and the two concepts model different things, and that the lowering of common stores/loads would be way more complicated than necessary in a design that mixes the two.

File: xdsl/dialects/vector.py
Comment (webmiche) [reply xdsl/dialects/vector.py:19]: Maybe a pytest makes more sense then? Or is there a reason why you suggested lit, @math-fehr?

File: tests/filecheck/vector_ops.xdsl
Comment (webmiche) [inline tests/filecheck/vector_ops.xdsl:5]: I prefer having very modular tests (though that certainly isn't the standard in the current test suite). Therefore, I would suggest to change this into a function that takes and returns a memref such that you can remove `Alloca`, `Dealloc` and `Constant`. In this way, you will not rely on the `arith`and `memref` dialects as much and there will be less conflicts if we decide to change something there.

File: tests/dialects/test_vector.py
Comment (webmiche) [inline tests/dialects/test_vector.py:30]: The `SSAValue` class actually is an abstract class, so it is not supposed to be instantiated directly. Unfortunately, Python does not give functionality to really disallow this from happening.

Please use `OpResult` instead, this should work the exact same way.

I know this is used in other tests, but it is a mistake there as well, just one that wasn't fixed yet.

File: tests/dialects/test_vector.py
Comment (webmiche) [inline tests/dialects/test_vector.py:14]: Please split these into three separate tests, one for eachvector you create. IMO, this will give better insights if tests fail.

**Review by webmiche — CHANGES_REQUESTED** [review]
LGTM generally, some style comments.

Also, I wanted to ask whether you have copied this modeling directly from MLIR and, if so, from which commit. We have a tagged commit in the README, which defines the exact version of all other dialects such that interoperation with MLIR works.

Secondly, just to be sure, the vector type does not have any costum printing in MLIR, right? 

Overall, I would like some interoperation with MLIR tests as well, if possible, in order to check that these things are in sync. Other mlir conversion tests can be found in `tests/filecheck/mlir-conversion`.

**Comment by meshtag** [comment]
> Also, I wanted to ask whether you have copied this modeling directly from MLIR and, if so, from which commit. We have a tagged commit in the README, which defines the exact version of all other dialects such that interoperation with MLIR works.

Yes, I have taken reference directly from `MLIR` and the generated `.mlir` code from `.xdsl` testing file works with commit hash: `74992f4a5bb79e2084abdef406ef2e5aa2024368` (the one which is mentioned in the project README). 

> Secondly, just to be sure, the vector type does not have any costum printing in MLIR, right?

Afaik, nope. 

> Overall, I would like some interoperation with MLIR tests as well, if possible, in order to check that these things are in sync.

I am not really sure if I understood this correctly (can you please elaborate). I have added a test [here](https://github.com/meshtag/xdsl/blob/vector_dialect/tests/filecheck/dialects/vector/vector_ops_mlir_conversion.xdsl) which compares generated `.mlir` code from the pure `.xdsl` example with expected IR. Does this address your suggestion? 

PS: I have created a separate `vector` dialect folder in `tests/filecheck` as suggested [here](https://github.com/xdslproject/xdsl/pull/282#discussion_r1057683792).

File: tests/filecheck/dialects/vector/vector_ops_mlir_conversion.xdsl
Comment (webmiche) [inline tests/filecheck/dialects/vector/vector_ops_mlir_conversion.xdsl:1]: I am fine with this approach as well, considering that you are not using any custom prints.

I was thinking of actually passing something to `mlir-opt`, then print and parse again with xDSL.

**Review by webmiche — APPROVED** [review]
LGTM, this can be merged after adding the tests Mathieu mentioned.

I suggest to just add the tests as pytests, I don't see any reason why filecheck is perferable.

**Comment by webmiche** [comment]
LGTM overall. I suggest to wait for @math-fehr s rereview and then merge. This might take some time (the days over Christmas/new year are always very slow here), but considering that there arent a lot of other PRs being pushed/merged, I feel we can get away with waiting.

File: xdsl/dialects/vector.py
Comment (math-fehr) [reply xdsl/dialects/vector.py:19]: For now, I would say that the best would be to have a file per error, but we would put them all in a folder to make this clean.
I think we should clearly have an `expected-error`, it's just that its implementation is a bit of a pain a think, and that we didn't need it that much before.
And the reason why I suggested `lit` is that I think we should always prioritize `lit` over `pytest`. I think that `lit` is a far better abstraction than `pytest` for these kind of tests, that do not rely on custom `Python` code.

**Review by math-fehr — APPROVED** [review]
LGTM, thanks a lot!

File: xdsl/dialects/vector.py
Comment (meshtag) [reply xdsl/dialects/vector.py:19]: Just to confirm, shall I create another PR to add `lit` tests for `vector.load/store` verifiers? 
This would contain 4 different files inside `tests/filecheck/dialects/vector` for each possible verifier exception.

Comment (math-fehr) [reply xdsl/dialects/vector.py:19]: I would say to me that `lit` makes more sense, though I'm not sure that everyone agrees!

**Pull request merged**
